lib/Transform/XTenMinimizeLiveTensors.cpp

```diff
@@ -0,0 +1,371 @@
+    for (auto &[op, info] : opToInfo) {
+      size_t outgoing = 0;
+      for_each(info.results,
+               [&outgoing](Value val) { outgoing += getSize(val); });
+      info.sizes.results = outgoing;
+      size_t incoming = 0;
+      for_each(info.operands,
+               [&incoming](Value val) { incoming += getSize(val); });
+      info.sizes.operands = incoming;
+      info.sizes.running = incoming;
+      if (!info.sharesResultMemory)
+        info.sizes.running += outgoing;
+    }
+  }
+
+  /// Recursively determine branch running sizes.
+  ///
+  /// \p opInfo points to the info for the op being analyzed.
+  /// \p brInfo incoming branch info, to be updated by this op.
+  /// \p completed contains the BranchRunning info for any fully
+  ///    analyzed operations.
+  void determineBranchRunning(OpInfo *opInfo, // NOLINT(misc-no-recursion)
+                              BranchRunning &brInfo,
+                              std::map<Operation *, BranchRunning> &completed) {
+    // Analyze simple fallthrough operations.
+    while (opInfo->operands.size() < 2 && opInfo->consumers.size() < 2) {
+      brInfo.maxRunning = std::max(brInfo.maxRunning, opInfo->sizes.results);
+      brInfo.lastResults = opInfo->sizes.results;
+      opInfo->orderedProducers = {brInfo.lastOp};
+      brInfo.lastOp = opInfo->op;
+      completed.insert({opInfo->op, brInfo});
+
+      if (opInfo->consumers.empty())
+        return; // Nothing more to compute.
+      opInfo = &opToInfo.at(opInfo->consumers.front());
+    }
+
+    // At a joining point - collect this branch and proceed iff all incoming
+    // branches have been collected.
+    SmallVector<BranchRunning> branches;
+    for (Value val : opInfo->operands) {
+      Operation *op = val.getDefiningOp();
+      if (op == nullptr)
+        op = currFn; // BlockArgument stand-in.
+
+      auto opIt = completed.find(op);
+      if (opIt == completed.end())
+        return; // Another producer needs to complete first.
+      branches.push_back(opIt->second);
+    }
+
+    std::sort(branches.begin(), branches.end(),
+              [](BranchRunning &aBranch, BranchRunning &bBranch) -> bool {
+                return (aBranch.maxRunning - aBranch.lastResults) >
+                       (bBranch.maxRunning - bBranch.lastResults);
+              });
+    opInfo->orderedProducers.clear();
+    for (BranchRunning const &branch : branches)
+      opInfo->orderedProducers.push_back(branch.lastOp);
+
+    // Complete the brInfo for this operation.
+    size_t maxRunning = opInfo->sizes.running;
+    for (BranchRunning const &branch : branches)
+      maxRunning = std::max(maxRunning, branch.maxRunning);
+    brInfo.maxRunning = maxRunning;
+    brInfo.lastResults = opInfo->sizes.results;
+    brInfo.lastOp = opInfo->op;
+    completed.insert({opInfo->op, brInfo});
+
+    // Continue to all consumers.
+    for (Operation *consumer : opInfo->consumers) {
+      BranchRunning nextRunning{.maxRunning = maxRunning,
```

Comment:
This is propagating too large a `maxRunning` into the innermost branching choices, were e.g. the `running` of the current `op` may dominate all consumer `running` and break the sorting within the inner branch-join graphs.

Reply:
@ehsan-toosi - I'll address this in a follow-up PR.